### PR TITLE
iniital commit of Java generator

### PR DESCRIPTION
**Pull request by rossiam**
- added simple Java template generator (at `smartthings:java`)
- moved node to sub-generator (`smartthings:node`)
- set up default generator ask for language first and proxy to appropriate sub-generator

### CODE REVIEW

File: generators/node/index.js [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+const yosay = require('yosay')
+const inquirer = require('inquirer')
+const Generator = require('../../lib/generator')
+const validator = require('./validator')
+
+module.exports = class extends Generator {
+  constructor(args, opts) {
+    super(args, opts)
+
+    this.option('skip-welcome', {
```
Comment (rossiam) [inline generators/node/index.js:13]: This option and its use below is the only actual change I made to this file (which was renamed from `generators/app.js`).

File: generators/app/index.js [hunk 2]
```diff
@@ -1,489 +1,30 @@
...
+
 const {default: chalk} = require('chalk')
 const yosay = require('yosay')
-const inquirer = require('inquirer')
-const Generator = require('../../lib/generator')
-const validator = require('./validator')
+const Generator = require('yeoman-generator')
 
 module.exports = class extends Generator {
-  constructor(args, opts) {
```
Comment (rossiam) [inline generators/app/index.js:10]: This file was moved and the name re-used so all the deleted lines here are in `generators/node/index.js`.

File: generators/java/index.js [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+    async function required(input) {
+      return input ? true : 'This field is required.'
+    }
+
+    Object.assign(this.appConfig, await this.prompt([{
+      type: 'input',
+      name: 'applicationName',
+      validate: required,
+      message: this._fmtMsg('What is the name your application in plain English (for documentation)?')
```
Comment (erodewald) [inline generators/java/index.js]: Not so sure about this phrasing – we don't require English. It's also a very long question that only needs to be about 4 words.

File: generators/java/templates/springboot/build.gradle.ejs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+dependencies {
+  implementation 'com.smartthings.sdk:smartapp-core:0.0.3-PREVIEW'
+  implementation 'com.smartthings.sdk:smartthings-client:0.0.3-PREVIEW'
+  implementation 'com.smartthings.sdk:smartapp-spring:0.0.3-PREVIEW'
+<% if (contextStore == 'dynamodb') { %>
+  implementation 'com.smartthings.sdk:smartapp-contextstore-dynamodb:0.0.3-PREVIEW'
+<% } %>
+	implementation 'org.springframework.boot:spring-boot-starter-web'
```
Comment (erodewald) [inline generators/java/templates/springboot/build.gradle.ejs]: I imagine this indentation appears in the output and looks funky.

File: generators/java/templates/springboot/build.gradle.ejs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  implementation 'ch.qos.logback:logback-classic:1.2.3'
+<% if (contextStore == 'dynamodb') { %>
+  implementation platform('com.amazonaws:aws-java-sdk-bom:1.11.568')
+  implementation 'com.amazonaws:aws-java-sdk-dynamodb'
+<% } %>
+
+	testImplementation 'org.springframework.boot:spring-boot-starter-test'
```
Comment (erodewald) [inline generators/java/templates/springboot/build.gradle.ejs]: Indentation again

File: generators/java/templates/springboot/src/main/java/__package_dir__/handlers/__class_name_prefix__ConfigurationHandler.java.ejs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        return Response.notFound();
+    }
+
+    private ExecutionResponse buildInitializeResponse() {
+        InitializeSetting initialize = new InitializeSetting()
+            .id("init")
+            .name("<%- applicationName %>")
+            .description("<%- applicationDescription %>")
+            .firstPageId("config")
+            .disableCustomDisplayName(false)
```
Comment (erodewald) [inline generators/java/templates/springboot/src/main/java/__package_dir__/handlers/__class_name_prefix__ConfigurationHandler.java.ejs:60]: Node version of this generator lets you choose `disableCustomDisplayName` and `disableRemoveApp`, why not match that?

File: generators/java/index.js [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        new inquirer.Separator(),
+        {name: 'None', value: 'none'}
+      ]
+    }))
+
+    if (this.appConfig.contextStore === 'dynamodb') {
+      Object.assign(this.appConfig, await this.prompt([{
+        type: 'input',
+        name: 'clientId',
+        message: this._fmtMsg('What is the client id for your app from the developer workspace?')
```
Comment (erodewald) [inline generators/java/index.js]: We should capitalize all instances of `developer workspace` as it is a proper noun. That probably needs to be changed in the NodeJS gen too but we'll start here.

File: generators/java/index.js (hunk 3 above)
Comment (rossiam) [reply generators/java/index.js]: Yeah, not sure why I put "English". What I was trying to get at was that this the regular name with spaces and such. Maybe just "What is the name your application for documentation?"?

File: generators/java/templates/springboot/build.gradle.ejs (hunk 4 above)
Comment (rossiam) [reply generators/java/templates/springboot/build.gradle.ejs]: Got a tab in accidentally...thought I had VSCode set to fix those. I'll update.

File: generators/java/templates/springboot/src/main/java/__package_dir__/handlers/__class_name_prefix__ConfigurationHandler.java.ejs (hunk 6 above)
Comment (rossiam) [reply generators/java/templates/springboot/src/main/java/__package_dir__/handlers/__class_name_prefix__ConfigurationHandler.java.ejs:60]: I can but I think we should be careful about asking too many questions for one-off things. I was using the rule of thumb that if it only needed to be changed one place and had a reasonable default, I would leave it out.

File: generators/java/index.js (hunk 7 above)
Comment (rossiam) [reply generators/java/index.js]: Gotcha, I'll update.

File: generators/java/index.js (hunk 3 above)
Comment (erodewald) [reply generators/java/index.js]: I would follow what Dev Workspace does: 

> This name will be used in the SmartThings app. Please provide a name that is friendly to the end user.

This can be condensed down to `Please provide a name that is friendly to the end user: ___`

File: generators/java/templates/springboot/src/main/java/__package_dir__/handlers/__class_name_prefix__ConfigurationHandler.java.ejs (hunk 6 above)
Comment (erodewald) [reply generators/java/templates/springboot/src/main/java/__package_dir__/handlers/__class_name_prefix__ConfigurationHandler.java.ejs:60]: That's why I buried those options behind a simple yes/no. I guess we can reevaluate and add later.

File: generators/java/index.js (hunk 3 above)
Comment (rossiam) [reply generators/java/index.js]: done 👍

**Review by erodewald — APPROVED** [review]
